triage/script.js

```diff
@@ -0,0 +1,738 @@
+};
+
+// Load options from form inputs, put them in the URL, and return the options dict.
+function readOptions() {
+	var read = id => {
+		let el = document.getElementById(id);
+		if (el.type === "checkbox") return el.checked;
+		if (el.type === "radio") return el.form[el.name].value;
+		if (el.type === "text") {
+			if (id.startsWith("filter")) {
+				if (el.value === "") {
+					return null;
+				}
+				try {
+					return new RegExp(el.value, "im");
+				} catch(err) {
+					console.error("bad regexp", el.value, err);
+					return new RegExp(RegExp.escape(el.value), "im");
+				}
+			} else {
+				return el.value;
+			}
+		}
+	}
+
+	var opts = {
+		ci: read('job-ci'),
+		pr: read('job-pr'),
+		reText: read('filter-text'),
+		reJob: read('filter-job'),
+		reTest: read('filter-test'),
+		showNormalize: read('show-normalize'),
+		sort: read('sort'),
+	}
+
+	var url = '';
+	if (!opts.ci) url += '&ci=0';
+	if (opts.pr) url += '&pr=1';
+	for (var name of ["text", "job", "test"]) {
+		var re = opts['re' + name[0].toUpperCase() + name.slice(1)];
+		if (re) {
+			var baseRe = re.toString().replace(/im$/, '').replace(/\\\//g, '/').slice(1, -1);
+			url += '&' + name + '=' + encodeURIComponent(baseRe);
+		}
+	}
+	if (url) {
+		if (document.location.hash) {
+			url += document.location.hash;
+		}
+		history.replaceState(null, "", "?" + url.slice(1));
+	} else if (document.location.search) {
+		history.replaceState(null, "", document.location.pathname + document.location.hash);
+	}
+
+	return opts;
+}
+
+// Convert querystring parameters into form inputs.
+function setOptionsFromURL() {
+	// http://stackoverflow.com/a/3855394/3694
+	var qs = (function(a) {
+		if (a == "") return {};
+		var b = {};
+		for (var i = 0; i < a.length; ++i)
+		{
+			var p=a[i].split('=', 2);
+			if (p.length == 1)
+				b[p[0]] = "";
+			else
+				b[p[0]] = decodeURIComponent(p[1].replace(/\+/g, " "));
+		}
+		return b;
+	})(window.location.search.substr(1).split('&'));
+
+	var write = (id, value) => {
+		if (!value) return;
+		var el = document.getElementById(id);
+		if (el.type === "checkbox") el.checked = (value === "1");
+		if (el.type === "text") el.value = value;
+	}
+	write('job-ci', qs.ci);
+	write('job-pr', qs.pr);
+	write('filter-text', qs.text);
+	write('filter-job', qs.job);
+	write('filter-test', qs.test);
+}
+
+// Render up to `count` clusters, with `start` being the first for consideration.
+function renderSubset(start, count) {
+	var top = document.getElementById('clusters');
+	var n = 0;
+	var shown = 0;
+	for (let [key, keyId, text, clusters] of clustered.c) {
+		if (n++ < start) continue;
+		shown += renderCluster(top, key, keyId, text, clusters);
+		lastClusterRendered = n;
+		if (shown >= count) break;
+	}
+}
+
+// Clear the page and reinitialize the renderer and filtering. Render a few failures.
+function rerender(maxCount) {
+	if (!clusteredAll) return;
+
+	options = readOptions();
+	clustered = clusteredAll.refilter(options);
```

Comment:
rerender() is called on first load and on form changes, so options would necessarily be different in this case.